support/reftest/reftest

```diff
@@ -0,0 +1,282 @@
+        help="""
+        prepare reference test data in dynamodb and s3
+        for integration test,
+         e.g. $./reftest prepare --bucket dichen-test --table dichen-test
+        """,
+    )
+    parser_prepare.add_argument(
+        "--release-date",
+        default=None,
+        help="Date on which the content will be made available.",
+    )
+    parse_aws_session(parser_prepare)
+    parse_cdn_certs(parser_prepare)
+    parser_prepare.add_argument(
+        "--bucket",
+        required=True,
+        help="The AWS S3 bucket used to store test data",
+    )
+    parser_prepare.add_argument(
+        "--table",
+        required=True,
+        help="The AWS dynamoDB used to store test data",
+    )
+
+    # parser for undo operation
+    parser_undo = subparsers.add_parser(
+        "undo",
+        help="""
+        the undo operation will clean reference test data in dynamodb,
+        but will not delete test data in s3,
+        e.g. $./reftest undo --table dichen-test
+        """,
+    )
+    parse_aws_session(parser_undo)
+    parser_undo.add_argument(
+        "--table",
+        required=True,
+        help="The AWS dynamoDB used to store test data",
+    )
+
+    return root_parser.parse_args()
+
+
+class RefTestConfig:
+    def __init__(self, prod_cdn_url, test_data):
+        self.prod_cdn_url = prod_cdn_url
+        self.test_data = test_data
+
+
+def load_config():
+    with open("data.yml") as f:
+        config = yaml.load(f, yaml.SafeLoader)
+    return RefTestConfig(config["prod-cdn-url"], config["test_data"],)
+
+
+def download_to_local(url, key_path, cert_path, cacert_path):
+    with requests.get(
+        url,
+        cert=(cert_path, key_path),
+        verify=cacert_path,
+        stream=True,
+        timeout=(30, 30),
+    ) as r:
+        r.raise_for_status()
+        total_size = int(r.headers.get("content-length", 0))
+        t = tqdm(desc=url, total=total_size, unit="iB", unit_scale=True)
+
+        # the file is deleted as soon as it is closed.
+        temp_file = tempfile.NamedTemporaryFile(delete=True)
+        for chunk in r.iter_content(chunk_size=8192):
+            if chunk:  # filter out keep-alive new chunks
+                t.update(len(chunk))
+                temp_file.write(chunk)
+                temp_file.flush()
+        t.close()
+        return temp_file
+
+
+def prepare(db, s3, config, opt):
+    for item in config.test_data:
+        url = config.prod_cdn_url + item["path"]
+        # download test data to local with a name "NamedTemporaryFile"
+        temp_file = download_to_local(url, opt.key, opt.cert, opt.cacert)
+        # push test data to s3 and dynamodb
+        s3.upload_from_localfile(
+            temp_file.name, item["sha256"], item["content-type"]
+        )
+        db.put_item(item["path"], item["sha256"], opt.release_date)
+        # delete the NamedTemporaryFile
+        temp_file.close()
+
+
+def undo(db, config):
+    for item in config.test_data:
```

Comment:
I still think the undo process here is not right.  This isn't limited to undoing whatever the command previously did, it's simply going to delete entries matching the paths and checksums of the testdata - whether or not they were put in place by this command.  And given that we picked real data at real paths, it's not at all unlikely that these will exist without having been put there by this script.  Listing files are the simplest example, e.g. if someone runs rhsm-update-listings against stage, it's likely to put /content/dist/rhel/server/5/5.7/listing in place with content exactly matching the reference test data.  This command shouldn't delete that entry just because it happens to match.

Can you please somehow make undo really just undo what the command did (e.g. by storing an extra piece of metadata flagging the entry as created by this command, as I mentioned earlier), or otherwise just remove the undo command?